crates/pipeline_manager/src/prober/mod.rs

```diff
@@ -0,0 +1,124 @@
+use crate::api::ManagerError;
+use crate::auth::TenantId;
+use crate::config::ProberConfig;
+use crate::db::storage::Storage;
+use crate::db::ProjectDB;
+use actix_web::{get, web, HttpServer, Responder};
+use log::{error, info};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::Mutex;
+use tokio::time::sleep;
+use tokio::{join, spawn};
+use uuid::Uuid;
+
+/// Task which indefinitely probes the services across all tenants.
+///
+/// TODO: improvements:
+/// - Fetch all current services for each tenant not just the default one
+/// - Parallel probing such that it is not done in sequence
+//  - Better waiting time between probes -- possibly exponential backoff?
+async fn probing_task(config: ProberConfig, db: Arc<Mutex<ProjectDB>>) -> Result<(), ManagerError> {
+    info!(
+        "Probing services started (sleeping {} seconds between each probing session)",
+        config.probe_sleep_inbetween_seconds
+    );
+
+    loop {
+        // Wait inbetween probes
+        sleep(Duration::from_secs(config.probe_sleep_inbetween_seconds)).await;
+
+        // Perform the probing
+        info!("Probing...");
+        let services = db
+            .lock()
+            .await
+            .list_services(TenantId(Uuid::nil()), &None)
+            .await;
+        match services {
+            Ok(services) => {
+                info!("Probing {} services...", services.len());
+                let mut success_counter = 0;
+                for service in &services {
+                    // info!("Probing service {}...", service.name);
+                    let started_at = chrono::offset::Utc::now();
+                    let result = service.config.probe(Duration::from_secs(10));
+                    let finished_at = chrono::offset::Utc::now();
+                    match db
+                        .lock()
+                        .await
+                        .new_service_probe(
```

Comment:
Will this just write to the DB forever?

Reply:
New version is only active rather than continuous in background.